intents/intent_ext.go

```diff
@@ -0,0 +1,231 @@
+		return fmt.Errorf("intent issued in the future")
+	}
+
+	// check if at least one signature is valid
+	if len(intent.Signers()) == 0 {
+		return fmt.Errorf("invalid signature")
+	}
+
+	// the intent is valid
+	return nil
+}
+
+func (intent *Intent) keyType(sessionId string) KeyType {
+	// handle empty session ids
+	if len(sessionId) == 0 {
+		return KeyTypeUnknown
+	}
+
+	// handle old session ids
+	if len(sessionId) == 42 {
+		return KeyTypeSECP256K1
+	}
+
+	// handle key typed session ids
+	sessionIdBytes := common.FromHex(sessionId)
+	return KeyType(sessionIdBytes[0])
+}
+
+func (intent *Intent) isValidSignature(sessionId string, signature string) error {
+	switch intent.keyType(sessionId) {
+	case KeyTypeSECP256K1:
+		return intent.isValidSignatureP256K1(sessionId, signature)
+	case KeyTypeSECP256R1:
+		return intent.isValidSignatureP256R1(sessionId, signature)
+	default:
+		return fmt.Errorf("unknown session key type")
+	}
+}
+
+// isValidSignatureP256K1 checks if the signature is valid for the given secp256k1 session
+func (intent *Intent) isValidSignatureP256K1(sessionId string, signature string) error {
+	// validate session address and signature
+	if !strings.HasPrefix(signature, "0x") || !strings.HasPrefix(sessionId, "0x") {
+		// invalid params
+		return fmt.Errorf("invalid sessionId or signature format")
+	}
+
+	// validate session address
+	sessionAddressBytes := common.FromHex(sessionId)
+	if len(sessionAddressBytes) != 21 && len(sessionAddressBytes) != 20 {
+		// invalid session address
+		return fmt.Errorf("invalid sessionId length")
+	}
+
+	// validate signature
+	sigBytes := common.FromHex(signature)
+	if len(sigBytes) != 65 {
+		return fmt.Errorf("invalid signature length")
+	}
+
+	// handle typed session address
+	if len(sessionAddressBytes) == 21 {
+		sessionAddressBytes = sessionAddressBytes[1:]
+	}
+
+	sessionAddress := fmt.Sprintf("0x%s", common.Bytes2Hex(sessionAddressBytes))
+
+	// Get hash of the packet
+	hash, err := intent.Hash()
+	if err != nil {
+		return fmt.Errorf("failed to hash intent: %w", err)
+	}
+
+	// Add Ethereum prefix to the hash
+	prefixedHash := crypto.Keccak256Hash([]byte(fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(hash), hash)))
+
+	// handle recovery byte
+	if sigBytes[64] == 27 || sigBytes[64] == 28 {
+		sigBytes[64] -= 27
+	}
+
+	// Recover the public key from the signature
+	pubKey, err := crypto.Ecrecover(prefixedHash.Bytes(), sigBytes)
+	if err != nil {
+		return fmt.Errorf("failed to recover public key: %w", err)
+	}
+
+	addr := common.BytesToAddress(crypto.Keccak256(pubKey[1:])[12:])
+
+	// Check if the recovered address matches the session address
+	if strings.ToLower(addr.Hex()) != strings.ToLower(sessionAddress) {
+		return fmt.Errorf("invalid signature")
+	}
+	return nil
+}
+
+// isValidSignatureP256R1 checks if the signature is valid for the given secp256r1 session
+func (intent *Intent) isValidSignatureP256R1(sessionId string, signature string) error {
```

Comment:
I would export this method, because there is no harm, and it could be useful by other callers in future etc

Reply:
Made it public and moved it to ``session.go``. Also added ``Sign`` functions in ``session.go`` for completeness.

This still can be executed using ``Intent.IsValid()``.